backend/src/chat/services/chatroom.socket.service.ts

```diff
@@ -0,0 +1,179 @@
+import { Injectable } from '@nestjs/common';
+import { WsException } from '@nestjs/websockets';
+import { ChatService } from '../../chat/chat.service';
+import { Socket, Server } from 'socket.io';
+import { ChatroomService } from '../../chatroom/chatroom.service';
+import { ChatroomUserService } from '../../chatroom_user/chatroom_user.service';
+import { ChatroomRestrictionService } from '../../chatroom_restriction/chatroom_restriction.service';
+import { RoomType } from '@prisma/client';
+import * as events from '../socketioEvents';
+import { CreateChatroomPayload } from '../interfaces/createChatroom.interface';
+
+@Injectable()
+export class ChatroomSocketService {
+  constructor(
+    private chat: ChatService,
+    private chatroom: ChatroomService,
+    private chatroomUser: ChatroomUserService,
+    private chatroomRestriction: ChatroomRestrictionService
+  ) {}
+
+  async createChatroom(client: Socket, payload: any) {
+    const userId = client['decoded'].sub;
+    const parsedPayload = this.parsePayload(payload);
+
+    this.validatePayload(parsedPayload);
+    return await this.chatroom.create(userId, parsedPayload);
+  }
+
+  async deleteChatroom(
+    client: Socket,
+    server: Server,
+    payload: { chatroomId: string }
+  ) {
+    const userId = client['decoded'].sub;
+    const chatroom = await this.chatroom.findOne(payload.chatroomId);
+
+    if (!(await this.chatroomUser.isUserOwner(userId, payload.chatroomId))) {
+      throw new WsException('Unauthorized to delete room');
+    }
+    this.chatroom.remove(chatroom.id);
+
+    server
+      .to(chatroom.slug)
+      .emit(events.CHATROOM_KICKED, { chatroomId: chatroom.id });
+    server.in(chatroom.slug).socketsLeave(chatroom.slug);
+  }
+
+  async joinChatroom(
+    client: Socket,
+    server: Server,
+    payload: { chatroomId: string; password: string }
+  ) {
+    const userId = client['decoded'].sub;
+    const chatroom = await this.chatroom.findOne(payload.chatroomId);
+
+    await this.chatroom.join(userId, payload.chatroomId, payload.password);
+
+    const formattedUser = await this.getFormattedChatroomUser(
+      userId,
+      chatroom.id
+    );
+
+    client.emit(events.CHATROOM_NEW, {
+      chatroom: chatroom
+    });
+    server.to(chatroom.slug).emit(events.CHATROOM_USER_CONNECT, {
+      chatroomId: chatroom.id,
+      chatroomUser: formattedUser
+    });
+  }
+
+  async leaveChatroom(
+    client: Socket,
+    server: Server,
+    payload: { chatroomId: string }
+  ) {
+    const userId = client['decoded'].sub;
+    const chatroom = await this.chatroom.findOne(payload.chatroomId);
+
+    if (chatroom.type !== RoomType.ONE_TO_ONE) {
```

Comment:
In front we can leave private message, is it working ?

Reply:
Done